File: inktui/src/bus/protocol.ts
/**
 * Transitional event/DTO compatibility types for store projections.
 *
 * The public application wire is generated at `../generated/applicationProtocol.ts`.
 * This older port remains only because projection payloads still reuse established
 * `BusEvent` DTOs while the backend implementation sits behind the application gateway.
 *
 * No transport may send the legacy message envelopes declared at the bottom of this file.
 *
 * Out of scope (rule 4): no sockets, no Ink, no framing loops, no transport. Types and constants
 * only — the same boundary the Python module draws ("if you find yourself importing asyncio here,
 * you're in the wrong file"). Framing/handshake logic lives in the `UdsBusClient` (C2).
 */

export const PROTOCOL_VERSION = 5;

// === Closed enums ============================================================
// Modeled as union types of string literals rather than TS `enum`s: the wire carries the bare
// string, `verbatimModuleSyntax`/`isolatedModules` discourage runtime `enum`, and a literal union
// is the idiomatic discriminant for the switch-heavy dispatch the store does. Adding a value here
// must bump PROTOCOL_VERSION, exactly as in the Python source.

export type Role =
  | 'collaborator'
  | 'notetaker'
  | 'planner'
  | 'planning_handler'
  | 'crow_handler'
  | 'crow';

export type AgentStatus =
  | 'idle'
  | 'running'
  | 'blocked'
  | 'escalating'
  | 'done'
  | 'failed'
  | 'dead';

export type CommandStatus = 'pending' | 'in_flight' | 'done' | 'failed' | 'cancelled';

/** Entity kind named by a key-only {@link StateSnapshotEvent}. Closed — adding a value bumps
 * PROTOCOL_VERSION. These are the slice names the store invalidates against.
 *
 * Mirrors the Python `Entity` enum (`murder/bus/protocol.py`) 1:1, including `report` (added to
 * Python in F1; the former C6 forward-declaration is now backed by the real backend enum). */
export type Entity =
  | 'ticket'
  | 'agent'
  | 'plan'
  | 'note'
  | 'report'
  | 'escalation'
  | 'queue_row'
  | 'history'
  | 'transit';

export type PresenceState = 'attended' | 'headless';

export type ClientKind = 'tui' | 'web' | 'cli_ephemeral' | 'worker';

// === Inner events (server -> view; discriminated by `type`) ===================
// Persisted to the service's events table. The store reacts to `state.snapshot` (slice
// invalidation) and `conversation.block` (transcript append); the rest are carried for
// completeness so the discriminated union is exhaustive and the compiler can prove a switch covers
// every kind (tsconfig `noFallthroughCasesInSwitch` / `noImplicitReturns`).

/** Fields common to every inner event. Mirrors Python `_BaseEvent`. `id`/`ts` are server-stamped
 * (UUID / ISO-8601 string on the wire); the optional fields are absent when the server omits
 * them. */
interface BaseEvent {
  id: string;
  ts: string;
  run_id: string;
  agent_id: string;
  role?: Role | null;
  ticket_id?: string | null;
}

export interface HeartbeatEvent extends BaseEvent {
  type: 'heartbeat';
  state: 'progressing' | 'stuck' | 'thinking';
  summary?: string | null;
  since_change_s: number;
}

export interface SummaryEvent extends BaseEvent {
  type: 'summary';
  text: string;
  checklist_done: number;
  checklist_total: number;
  last_message_excerpt: string;
}

export interface QuestionEvent extends BaseEvent {
  type: 'question';
  question: string;
  crow_session: string;
  recent_pane: string;
}

/** A free-text working note emitted by a crow via a `>>> NOTE:` marker. Under DB-owns-runtime
 * these land in the events table (audit log), not the ticket `.md`. Carried for union
 * completeness; the store re-pulls the notes slice via the `note` {@link Entity} `state.snapshot`,
 * so this event needs no dedicated handler. Mirrors Python `NoteEvent`. */
export interface NoteEvent extends BaseEvent {
  type: 'note';
  note: string;
}

export interface EscalationEvent extends BaseEvent {
  type: 'escalation';
  to: 'user' | 'collaborator';
  reason: string;
  severity: 1 | 2 | 3;
  crow_session?: string | null;
  source_event_id?: string | null;
}

export interface StatusChangeEvent extends BaseEvent {
  type: 'status_change';
  entity: 'agent' | 'ticket';
  entity_id: string;
  from_status: string;
  to_status: string;
  reason?: string | null;
}

export interface ErrorEvent extends BaseEvent {
  type: 'error';
  message: string;
  recoverable: boolean;
  traceback?: string | null;
}

export interface CommandEvent extends BaseEvent {
  type: 'command';
  target_worker: string;
  kind: string;
  payload: Record<string, unknown>;
  correlation_id: string;
  idempotency_key: string;
  status: CommandStatus;
  claimed_by?: string | null;
  lease_expires_at?: number | null;
  attempt_count: number;
  retryable: boolean;
  result?: Record<string, unknown> | null;
}

/** Key-only notification that an entity changed. The body lives in the service; the store re-pulls
 * the named slice on receipt (the perf story that replaced poll-everything-every-tick). This is
 * the event the store subscribes to for slice invalidation. */
export interface StateSnapshotEvent extends BaseEvent {
  type: 'state.snapshot';
  entity: Entity;
  key: string;
  entity_version: number;
  /** Reserved forward-compat field (F1, unused now). The contract stays key-only: clients refetch
   * the named slice. A future low-bandwidth mode MAY inline the changed data here to skip the
   * refetch round-trip. Absent on the wire today; do not build a translation layer off it. */
  payload?: Record<string, unknown> | null;
}

export interface PresenceEvent extends BaseEvent {
  type: 'presence';
  state: PresenceState;
  user_count: number;
  /** Per-{@link ClientKind} connection counts, keyed by the kind string. */
  kinds: Record<string, number>;
  /** Monotonic per supervisor lifetime; subscribers ignore non-increasing versions. */
  version: number;
}

export interface SchedulerModeEvent extends BaseEvent {
  type: 'scheduler.mode';
  from_mode: string;
  to_mode: string;
  changed_by: 'user' | 'api';
}

export interface SchedulerDecisionEvent extends BaseEvent {
  type: 'scheduler.decision';
  mode: string;
  harness: string;
  window_key: string;
  decision: boolean;
  usage: number;
  t_until_reset: number;
  t_period: number;
  threshold: number;
  rationale: string;
  kicked_ticket_id?: string | null;
}

/** Completion coordinator verdict for a ticket (peer of {@link SchedulerDecisionEvent}). Mirrors
 * `CompletionVerdictEvent` in `murder/bus/protocol.py`. SERVER-SIDE FORENSIC event: the client does
 * NOT act on it directly — it reads completion via the key-only `state.snapshot` + slice refetch
 * path. Declared here only so the discriminated union parses it. */
export interface CompletionVerdictEvent extends BaseEvent {
  type: 'completion.verdict';
  completed: boolean;
  ticket_failed: boolean;
  failed_checks: string[];
}

/** Rich agent-registry mutation (register / rename / clear / force_stop). Mirrors
 * `AgentLifecycleEvent` in `murder/bus/protocol.py`. SERVER-SIDE FORENSIC event: the client does NOT
 * act on it directly — agent state arrives via the key-only `state.snapshot` path. Declared here
 * only so the discriminated union parses it. */
export interface AgentLifecycleEvent extends BaseEvent {
  type: 'agent.lifecycle';
  op: 'register' | 'rename' | 'clear' | 'force_stop';
  details: Record<string, unknown>;
  reason?: string | null;
}

export interface UsageResetEvent extends BaseEvent {
  type: 'usage.reset';
  harness: string;
  prev_pct: number;
  curr_pct: number;
}

/** Content-bearing conversation block. `action` distinguishes an immutable append from a live
 * trailing-block update. The store appends/replaces the block in the named conversation's
 * transcript slice. `block` is left opaque here — its shape is the transcript DTO, owned above the
 * transport seam.
 *
 * TUIchat-4 (Condensed view) reuses THIS event channel for a third `action`, `chunk-summarized`:
 * the Python producer (`ConversationProducer._summarize_chunk` →
 * `AgentBase._publish_conversation_block`) wraps every per-flush summary in a `conversation.block`
 * event with `action: 'chunk-summarized'` — NOT a distinct top-level event `type`. For that action
 * `block` is the chunk-summary payload `{ conversation_id, summary, block_ids }` (see
 * {@link ConversationChunkSummaryEventBlock}), NOT a transcript-block row. The conversations slice
 * folds it into its ephemeral `chunkSummaries` map for incremental Condensed updates between
 * snapshots; the `state.conversations_snapshot` chunk_summaries[] remains the source of truth. */
export interface ConversationBlockEvent extends BaseEvent {
  type: 'conversation.block';
  conversation_id: string;
  action: 'block-appended' | 'block-updated' | 'chunk-summarized';
  block: Record<string, unknown>;
}

/** The `block` payload of a `conversation.block` event whose `action` is `'chunk-summarized'`
 * (TUIchat-4). Snake_case to match the Python wire (`conversation_producer.py` publishes
 * `{conversation_id, summary, block_ids}`). `summary` stands in for exactly the blocks named in
 * `block_ids` in the Condensed view. The snapshot omits `summary_id`/`chunk_idx` here (those come
 * with the authoritative `chunk_summaries[]` snapshot rows); the live event is an incremental hint. */
export interface ConversationChunkSummaryEventBlock {
  conversation_id: string;
  summary: string;
  block_ids: readonly number[];
}

/** Content-bearing conversation liveness push. Companion to {@link ConversationBlockEvent} on the
 * same additive-event-kind seam (mirrors `ConversationStateEvent` in `murder/bus/protocol.py`):
 * emitted whenever a conversation's parsed harness UI state (`working` / `awaiting_input` /
 * `awaiting_approval`) or its queued-but-undelivered user message changes. The conversations slice
 * stores both per agent so the chat input can render the queued line + awaiting badge live. */
export interface ConversationStateEvent extends BaseEvent {
  type: 'conversation.state';
  conversation_id: string;
  /** Parsed harness UI state at last projection: working | awaiting_input | awaiting_approval. */
  live_state: string | null;
  /** A user message accepted while the harness was busy, held for idle delivery; null when none. */
  queued_message: string | null;
}

/**
 * Raw ANSI frame from tmux for the focused pane. The CLIENT subscribes on `ctrl+y` enter and
 * disposes on exit. CAVEAT — the disposer is LOCAL-only: the wire protocol has no `unsub` op, so
 * the server keeps streaming frames over the multiplexed connection until the whole connection
 * closes; the client simply stops fanning them out (see `UdsBusClient.subscribe`'s disposer). So
 * this is NOT a zero standing cost after exit — the frames keep arriving and are dropped client-
 * side. Closing that gap for real needs a server-side per-subscription teardown op (a `unsub`
 * frame the server honours), which the bus does not yet expose. Track that as the real fix.
 *
 * Mirrors `TmuxFrameEvent` in `murder/bus/protocol.py` — both added in F6, `PROTOCOL_VERSION`
 * bumped to 3 in lockstep across both files.
 *
 * The frame carries the full rendered ANSI output of the pane as a snapshot string (from
 * `tmux capture-pane -e`). The consumer replaces its display on every event (no incremental
 * patching). Ink `<Text>` renders ANSI escape sequences natively.
 *
 * Pane-scoping: subscribe with `agent_id` in the {@link EventFilter} to stream that agent's own
 * tmux session; without it the service falls back to its project session.
 */
export interface TmuxFrameEvent extends BaseEvent {
  type: 'tmux.frame';
  /** The full rendered ANSI content for the focused tmux pane (snapshot, not incremental). */
  frame: string;
}

/** A durable external-decision request projected from verified harness evidence. */
export interface HarnessDecisionRequestEvent extends BaseEvent {
  type: 'harness.decision.request';
  decision_request_id: string;
  decision_kind: 'question' | 'permission';
  request_identity: string;
  observation_revision: readonly [number, number, number];
  request: Record<string, unknown>;
}

/** A recorded user/policy choice, persisted before verified execution. */
export interface HarnessDecisionResponseEvent extends BaseEvent {
  type: 'harness.decision.response';
  decision_request_id: string;
  decision_kind: 'question' | 'permission';
  request_identity: string;
  response: Record<string, unknown>;
  decided_by: string;
}

/** Discriminated union of every server-pushed inner event. `type` is the discriminant. */
export type BusEvent =
  | HeartbeatEvent
  | SummaryEvent
  | QuestionEvent
  | NoteEvent
  | EscalationEvent
  | StatusChangeEvent
  | ErrorEvent
  | CommandEvent
  | StateSnapshotEvent
  | PresenceEvent
  | SchedulerModeEvent
  | SchedulerDecisionEvent
  | CompletionVerdictEvent
  | AgentLifecycleEvent
  | UsageResetEvent
  | ConversationBlockEvent
  | ConversationStateEvent
  | TmuxFrameEvent
  | HarnessDecisionRequestEvent
  | HarnessDecisionResponseEvent;

/** The string literal `type` discriminant of every {@link BusEvent}. */
export type BusEventType = BusEvent['type'];

// === Filter ==================================================================

/** Server-applied subscription filter. Fields compose with AND; an absent field matches any. The
 * broker applies this before fanout. Mirrors Python `EventFilter`. */
export interface EventFilter {
  role?: Role;
  ticket_id?: string;
  type?: BusEventType;
  entity?: Entity;
  target_worker?: string;
  kind?: string;
  agent_id?: string;
}

// === Envelope bodies =========================================================

/** First message a client sends after connect. The server replies with an {@link AckMessage}
 * (`kind: 'subscribed'`) on success or an {@link ErrMessage} (`code: 'protocol_version_mismatch'`)
 * on version disagreement. `client_id` is stable across reconnects so the supervisor can resume
 * RPC/presence state. */
export interface HelloBody {
  protocol_version: number;
  client_kind: ClientKind;
  client_id: string;
  since_id?: number | null;
}

/** Subscribe arguments. `since_id=N` replays every persisted event with id > N before the live
 * tail, terminated by an {@link AckMessage} (`kind: 'replay_done'`). `tail_only` skips replay and
 * delivers only events published after subscribe (the server uses the current watermark as the
 * replay cursor). `presence_retain` makes the current presence sticky-retained to late subscribers. */
export interface SubArgs {
  filter: EventFilter;
  since_id?: number | null;
  tail_only?: boolean;
  presence_retain: boolean;
}

/** Topic names accepted by the `hydrate` wire op. `all` is the server-defined convenience topic;
 * concrete topic strings are intentionally open so new views can hydrate narrower state without
 * changing this transport layer. */
export type HydrateTopic = 'all' | (string & {});

/** Hydrate arguments. The client owns `cursor`: application code never passes it directly. */
export interface HydrateArgs {
  topics: readonly HydrateTopic[];
  cursor?: number | null;
}

/** RPC arguments. `target` is the method name; `body` its params. Mirrors Python `RpcArgs`. */
export interface RpcArgs {
  target: string;
  body: Record<string, unknown>;
  timeout_s: number;
}

export interface AckBody {
  kind: 'subscribed' | 'replay_done' | 'rpc_reply' | 'hydrate_reply' | 'pong' | 'published';
  watermark?: number | null;
  result?: Record<string, unknown> | null;
}

export interface ErrBody {
  code: string;
  message: string;
  /** Optional: Python `ErrBody.details` defaults to `{}` (`Field(default_factory=dict)`) and is
   * omitted on the wire when empty, so the client must not require it (C-D3). */
  details?: Record<string, unknown>;
}

/** Per-client signal on connect — distinct from the broadcast {@link PresenceEvent}.
 * Names the entities whose state is most likely stale for the joining client. Not persisted. */
export interface WakeBody {
  client_id: string;
  /** Only `'connect'` is ever sent today: the Python server constructs `WakeBody` with
   * `reason="connect"` in exactly one place (`transport_socket.py`) and never `"reconnect"`. The
   * Python type keeps `"reconnect"` as a forward-compat Literal, but the consumer models only what
   * is actually emitted (C-LM-5). Re-add `'reconnect'` here if the server starts sending it. */
  reason: 'connect';
  fresh_state_hints: Entity[];
}

// === Wire envelope (client <-> service; discriminated by `op`) ================

interface BaseMessage {
  schema_version: number;
  correlation_id: string;
}

export interface HelloMessage extends BaseMessage {
  op: 'hello';
  body: HelloBody;
}

export interface PubMessage extends BaseMessage {
  op: 'pub';
  /** Durable event-log position for this frame. Optional while older servers are in flight. */
  seq?: number | null;
  event: BusEvent;
}

export interface SubMessage extends BaseMessage {
  op: 'sub';
  args: SubArgs;
}

export interface RpcMessage extends BaseMessage {
  op: 'rpc';
  args: RpcArgs;
}

export interface HydrateMessage extends BaseMessage {
  op: 'hydrate';
  args: HydrateArgs;
}

export interface AckMessage extends BaseMessage {
  op: 'ack';
  body: AckBody;
}

export interface ErrMessage extends BaseMessage {
  op: 'err';
  body: ErrBody;
}

export interface WakeMessage extends BaseMessage {
  op: 'wake';
  body: WakeBody;
}

/** Discriminated union of every wire envelope. `op` is the discriminant. JSON-lines framed on the
 * socket (one envelope per `\n`-terminated line); the framing itself lives in C2. */
export type WireMessage =
  | HelloMessage
  | PubMessage
  | SubMessage
  | RpcMessage
  | HydrateMessage
  | AckMessage
  | ErrMessage
  | WakeMessage;

/** The string literal `op` discriminant of every {@link WireMessage}. */
export type WireOp = WireMessage['op'];

// === Wire constants ==========================================================
// Mirrors the Python `=== Wire constants ===` block. C2 resolves the socket path from the runtime
// subdir + basename; the timeouts seed the real client's RPC/handshake deadlines.

export const SOCKET_RUNTIME_SUBDIR = 'murder';
export const SOCKET_BASENAME = 'bus.sock';

export const DEFAULT_RPC_TIMEOUT_S = 30.0;
export const DEFAULT_HEARTBEAT_INTERVAL_S = 5.0;
export const DEFAULT_LEASE_TTL_S = 30.0;
export const DEFAULT_MAX_COMMAND_ATTEMPTS = 3;
export const COMMAND_REAPER_INTERVAL_S = 5.0;

export const PRESENCE_DISCONNECT_DEBOUNCE_S = 30.0;
/** Only these client kinds count toward {@link PresenceEvent}.user_count. */
export const PRESENCE_USER_KINDS: readonly ClientKind[] = ['tui', 'web'];

export const SUBSCRIBER_QUEUE_DEFAULT = 1024;
export const IDEMPOTENCY_WINDOW_S = 60.0;

File: murder/bus/protocol.py
"""Frozen wire contract between supervisor, workers, and clients.

This module is the single source of truth for the bus protocol that spans
the worker-bus refactor. Both the TUI branch and the backend branch import
from here. The TUI develops against a fake bus that produces messages
matching these types; the backend builds the broker that produces them
for real. They meet at this file.

Amendment process
-----------------
- **Additive changes** (new optional fields, new BusEvent kinds, new
  closed-enum values that don't change existing meaning) MAY be made
  inline. Coordinate the PR across both branches.
- **Non-additive changes** (renaming a field, changing a discriminator
  value, narrowing a type, removing a kind) MUST bump
  ``PROTOCOL_VERSION``. Clients refuse mismatched versions on connect.

Out of scope
------------
Broker implementations, transport framing read/write loops, kind handler
bodies, asyncio glue, sqlite glue. Types and constants only. If you find
yourself importing ``asyncio`` here, you're in the wrong file.

See `.agents/bus_protocol.md` for the design rationale behind these
shapes.
"""

from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Annotated, Any, Literal
from uuid import UUID, uuid4

try:
    from enum import StrEnum
except ImportError:  # Python <3.11

    class StrEnum(str, Enum):
        def __str__(self) -> str:
            return str.__str__(self)


from pydantic import BaseModel, Field, TypeAdapter

from murder.work.tickets.status import TicketStatus

PROTOCOL_VERSION = 3


# === Closed enums ============================================================


class Role(StrEnum):
    COLLABORATOR = "collaborator"
    NOTETAKER = "notetaker"
    PLANNER = "planner"
    PLANNING_HANDLER = "planning_handler"
    CROW_HANDLER = "crow_handler"
    CROW = "crow"


class AgentStatus(StrEnum):
    IDLE = "idle"
    RUNNING = "running"
    BLOCKED = "blocked"
    ESCALATING = "escalating"
    DONE = "done"
    FAILED = "failed"
    DEAD = "dead"


class CommandStatus(StrEnum):
    PENDING = "pending"
    IN_FLIGHT = "in_flight"
    DONE = "done"
    FAILED = "failed"
    CANCELLED = "cancelled"


class Entity(StrEnum):
    """Closed — adding a value bumps PROTOCOL_VERSION."""

    TICKET = "ticket"
    AGENT = "agent"
    PLAN = "plan"
    NOTE = "note"
    REPORT = "report"
    ESCALATION = "escalation"
    QUEUE_ROW = "queue_row"


class PresenceState(StrEnum):
    ATTENDED = "attended"
    HEADLESS = "headless"


class ClientKind(StrEnum):
    TUI = "tui"
    WEB = "web"
    CLI_EPHEMERAL = "cli_ephemeral"
    WORKER = "worker"


# === Inner events (persisted to events table; discriminated by `type`) =======


class _BaseEvent(BaseModel):
    id: UUID = Field(default_factory=uuid4)
    ts: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    run_id: str
    agent_id: str = ""
    role: Role | None = None
    ticket_id: str | None = None


# --- Legacy types (semantics unchanged from murder/bus.py pre-refactor) ------


class HeartbeatEvent(_BaseEvent):
    type: Literal["heartbeat"] = "heartbeat"
    state: Literal["progressing", "stuck", "thinking"]
    summary: str | None = None
    since_change_s: int = 0


class SummaryEvent(_BaseEvent):
    type: Literal["summary"] = "summary"
    text: str
    checklist_done: int = 0
    checklist_total: int = 0
    last_message_excerpt: str = ""


class QuestionEvent(_BaseEvent):
    type: Literal["question"] = "question"
    question: str
    crow_session: str
    recent_pane: str = ""


class NoteEvent(_BaseEvent):
    """A free-text working note emitted by a crow via a ``>>> NOTE:`` marker.

    Under DB-owns-runtime these land in the ``events`` table (audit log), not
    the ticket ``.md`` — the bus persists every event before fan-out.
    """

    type: Literal["note"] = "note"
    note: str


class EscalationEvent(_BaseEvent):
    type: Literal["escalation"] = "escalation"
    to: Literal["user", "collaborator"]
    reason: str
    severity: Literal[1, 2, 3] = 2
    crow_session: str | None = None
    source_event_id: UUID | None = None


class StatusChangeEvent(_BaseEvent):
    type: Literal["status_change"] = "status_change"
    entity: Literal["agent", "ticket"]
    entity_id: str
    from_status: str
    to_status: str
    reason: str | None = None


class ErrorEvent(_BaseEvent):
    type: Literal["error"] = "error"
    message: str
    recoverable: bool = True
    traceback: str | None = None


# --- New types added by the worker-bus refactor -----------------------------


class CommandEvent(_BaseEvent):
    """A request to a worker. Lives in both ``commands`` and ``events``.

    The ``commands`` row is the work queue; the ``events`` row is the
    audit log. Both are written in one transaction. ``events.payload_json``
    carries enough to reconstruct the command id; the full body lives in
    ``commands``.
    """

    type: Literal["command"] = "command"
    target_worker: str
    kind: str
    payload: dict[str, Any] = Field(default_factory=dict)
    correlation_id: str
    idempotency_key: str
    status: CommandStatus = CommandStatus.PENDING
    claimed_by: str | None = None
    lease_expires_at: int | None = None
    attempt_count: int = 0
    retryable: bool = True
    result: dict[str, Any] | None = None


class StateSnapshotEvent(_BaseEvent):
    """Key-only notification that an entity changed.

    Body lives in SQLite; every client holds a read-only DB handle and
    refetches on receipt. See worker_bus_refactor.md §D.5 for the
    deliberation.
    """

    type: Literal["state.snapshot"] = "state.snapshot"
    entity: Entity
    key: str
    entity_version: int = 0
    # Reserved forward-compat field (F1, unused now). The bus->client contract
    # stays key-only: default behaviour is a tiny key-only push and the client
    # refetches the named slice. A future low-bandwidth mode MAY inline the
    # changed data here (paired with field-masked snapshot RPCs) to skip the
    # refetch round-trip. This is a superset of what clients consume today --
    # no second event format, no migration. Do NOT build a translation layer
    # off this field as part of F1.
    payload: dict[str, Any] | None = None


class PresenceEvent(_BaseEvent):
    """Broadcast when the supervisor's debounced presence state transitions.

    Asymmetric debounce: connect→attended is immediate (responsive ramp);
    disconnect→headless is delayed by PRESENCE_DISCONNECT_DEBOUNCE_S to
    absorb blips.

    ``version`` is monotonic per supervisor lifetime. Subscribers MUST
    ignore events with version not strictly greater than the last seen,
    so that bridges that don't preserve order can't corrupt downstream
    workers' adaptive cadence.
    """

    type: Literal["presence"] = "presence"
    state: PresenceState
    user_count: int = 0
    kinds: dict[str, int] = Field(default_factory=dict)
    version: int


class SchedulerModeEvent(_BaseEvent):
    """Broadcast when the scheduler mode changes."""

    type: Literal["scheduler.mode"] = "scheduler.mode"
    from_mode: str
    to_mode: str
    changed_by: Literal["user", "api"] = "user"


class SchedulerDecisionEvent(_BaseEvent):
    """Emitted on every crow_magic tick per (harness, window_key)."""

    type: Literal["scheduler.decision"] = "scheduler.decision"
    mode: str
    harness: str
    window_key: str
    decision: bool
    usage: float
    t_until_reset: float
    t_period: float
    threshold: float
    rationale: str
    kicked_ticket_id: str | None = None


class UsageResetEvent(_BaseEvent):
    """Emitted when consecutive snapshots show a usage reset (drop > 80% of prev)."""

    type: Literal["usage.reset"] = "usage.reset"
    harness: str
    prev_pct: float
    curr_pct: float


class ConversationBlockEvent(_BaseEvent):
    """Content-bearing conversation block push event.

    Additive event kind for event-sourced transcripts. ``action`` distinguishes
    immutable appends from live trailing-block updates without touching the
    closed ``Entity`` enum or key-only ``StateSnapshotEvent`` contract.
    """

    type: Literal["conversation.block"] = "conversation.block"
    conversation_id: str
    action: Literal["block-appended", "block-updated"]
    block: dict[str, Any]


class ConversationStateEvent(_BaseEvent):
    """Content-bearing conversation liveness push event.

    Companion to ``ConversationBlockEvent`` on the same additive-event-kind
    seam: emitted whenever a conversation's parsed harness UI state
    (``working`` / ``awaiting_input`` / ``awaiting_approval``) or its queued
    -but-undelivered user message changes. Key-only ``StateSnapshotEvent``
    semantics do not fit here: the TUI renders the queued line / awaiting
    badge live and there is no snapshot entity for conversations.
    """

    type: Literal["conversation.state"] = "conversation.state"
    conversation_id: str
    live_state: str | None = None
    queued_message: str | None = None


class TmuxFrameEvent(_BaseEvent):
    """Raw ANSI frame from tmux for the focused pane.

    Streamed **only** while the tmux fullscreen mode is active (``ctrl+y``
    in the Ink TUI); the bus subscription is opened on enter and closed on
    exit. There is **no standing cost** when nobody is subscribed — the
    capture loop runs only for the lifetime of the subscription task.

    The ``frame`` field carries the full rendered ANSI output of the pane
    as a single snapshot string (from ``tmux capture-pane -e``). The
    consumer replaces its display on every event (no incremental patching).

    Pane-scoping note: the current ``EventFilter`` has no ``pane_id``
    field.  The server delivers a single stream for the configured focused
    pane; per-pane multiplexing is deferred (flagged in ``protocol.ts``
    C14 note).
    """

    type: Literal["tmux.frame"] = "tmux.frame"
    frame: str


# --- Discriminated union of inner events ------------------------------------

BusEvent = Annotated[
    HeartbeatEvent
    | SummaryEvent
    | QuestionEvent
    | NoteEvent
    | EscalationEvent
    | StatusChangeEvent
    | ErrorEvent
    | CommandEvent
    | StateSnapshotEvent
    | PresenceEvent
    | SchedulerModeEvent
    | SchedulerDecisionEvent
    | UsageResetEvent
    | ConversationBlockEvent
    | ConversationStateEvent
    | TmuxFrameEvent,
    Field(discriminator="type"),
]

BUS_EVENT_ADAPTER: TypeAdapter[BusEvent] = TypeAdapter(BusEvent)


# Back-compat: the legacy ``AgentEvent`` excludes the three new types so
# call sites that pattern-match on it keep their narrowed semantics until
# they migrate to ``BusEvent``.
AgentEvent = Annotated[
    HeartbeatEvent
    | SummaryEvent
    | QuestionEvent
    | EscalationEvent
    | StatusChangeEvent
    | ErrorEvent,
    Field(discriminator="type"),
]


# === Filter ==================================================================


class EventFilter(BaseModel):
    """Server-applied filter. Fields compose with AND; ``None`` matches any.

    The broker MUST apply filters before fanout. At the future tier-3 scale
    (~45 clients in worker_bus_refactor handoff §5.2) client-side filtering
    is O(clients × events) and untenable.
    """

    role: Role | None = None
    ticket_id: str | None = None
    type: str | None = None
    entity: Entity | None = None
    target_worker: str | None = None
    kind: str | None = None
    agent_id: str | None = None

    def matches(self, event: BaseModel) -> bool:
        if self.role is not None and getattr(event, "role", None) != self.role:
            return False
        if self.agent_id is not None and getattr(event, "agent_id", None) != self.agent_id:
            return False
        if self.ticket_id is not None and getattr(event, "ticket_id", None) != self.ticket_id:
            return False
        if self.type is not None and getattr(event, "type", None) != self.type:
            return False
        if self.entity is not None and getattr(event, "entity", None) != self.entity:
            return False
        if (
            self.target_worker is not None
            and getattr(event, "target_worker", None) != self.target_worker
        ):
            return False
        if self.kind is not None and getattr(event, "kind", None) != self.kind:
            return False
        return True


# === Envelope bodies =========================================================


class HelloBody(BaseModel):
    """Sent by the client as the first message after socket connect.

    Server responds with ``AckBody(kind="subscribed")`` after a successful
    handshake or ``ErrBody(code="protocol_version_mismatch")`` if versions
    disagree. ``client_id`` is stable across reconnects so the supervisor
    can resume in-flight RPC state and presence counting correctly.
    """

    protocol_version: int
    client_kind: ClientKind
    client_id: str
    since_id: int | None = None


class SubArgs(BaseModel):
    """Subscribe arguments. ``since_id=N`` MUST replay every persisted event
    with ``events.id > N`` before the live tail begins, terminated by
    ``AckBody(kind="replay_done", watermark=...)``.

    ``presence_retain=True`` makes presence sticky-retain: server emits the
    current PresenceEvent on the new subscription immediately after
    replay_done so late subscribers aren't blind to current state.
    """

    filter: EventFilter = Field(default_factory=EventFilter)
    since_id: int | None = None
    presence_retain: bool = True


class RpcArgs(BaseModel):
    target: str
    body: dict[str, Any] = Field(default_factory=dict)
    timeout_s: float = 30.0


class AckBody(BaseModel):
    kind: Literal["subscribed", "replay_done", "rpc_reply", "pong"]
    watermark: int | None = None
    result: dict[str, Any] | None = None


class ErrBody(BaseModel):
    code: str
    message: str
    details: dict[str, Any] = Field(default_factory=dict)


class WakeBody(BaseModel):
    """Per-client signal sent immediately on connect / reconnect.

    Distinct from PresenceEvent. PresenceEvent is broadcast and represents
    a global debounced transition; WakeEvent is per-client and represents
    "you just joined — here are the entities whose state is most likely
    stale for you." Not persisted to ``events``.
    """

    client_id: str
    reason: Literal["connect", "reconnect"] = "connect"
    fresh_state_hints: list[Entity] = Field(default_factory=list)


# === Wire envelope (discriminated by `op`) ===================================


class _BaseMessage(BaseModel):
    schema_version: int = PROTOCOL_VERSION
    correlation_id: str = ""


class HelloMessage(_BaseMessage):
    op: Literal["hello"] = "hello"
    body: HelloBody


class PubMessage(_BaseMessage):
    op: Literal["pub"] = "pub"
    event: BusEvent


class SubMessage(_BaseMessage):
    op: Literal["sub"] = "sub"
    args: SubArgs


class RpcMessage(_BaseMessage):
    op: Literal["rpc"] = "rpc"
    args: RpcArgs


class AckMessage(_BaseMessage):
    op: Literal["ack"] = "ack"
    body: AckBody


class ErrMessage(_BaseMessage):
    op: Literal["err"] = "err"
    body: ErrBody


class WakeMessage(_BaseMessage):
    op: Literal["wake"] = "wake"
    body: WakeBody


WireMessage = Annotated[
    HelloMessage | PubMessage | SubMessage | RpcMessage | AckMessage | ErrMessage | WakeMessage,
    Field(discriminator="op"),
]

WIRE_MESSAGE_ADAPTER: TypeAdapter[WireMessage] = TypeAdapter(WireMessage)


# === Wire constants ==========================================================

SOCKET_RUNTIME_SUBDIR = "murder"
SOCKET_BASENAME = "bus.sock"

DEFAULT_RPC_TIMEOUT_S = 30.0
DEFAULT_HEARTBEAT_INTERVAL_S = 5.0
DEFAULT_LEASE_TTL_S = 30.0
DEFAULT_MAX_COMMAND_ATTEMPTS = 3
COMMAND_REAPER_INTERVAL_S = 5.0

PRESENCE_DISCONNECT_DEBOUNCE_S = 30.0
PRESENCE_USER_KINDS: frozenset[ClientKind] = frozenset({ClientKind.TUI, ClientKind.WEB})
# Only these client kinds count toward PresenceEvent.user_count.

SUBSCRIBER_QUEUE_DEFAULT = 1024
IDEMPOTENCY_WINDOW_S = 60.0


__all__ = [
    "PROTOCOL_VERSION",
    "Role",
    "TicketStatus",
    "AgentStatus",
    "CommandStatus",
    "Entity",
    "PresenceState",
    "ClientKind",
    "HeartbeatEvent",
    "SummaryEvent",
    "QuestionEvent",
    "NoteEvent",
    "EscalationEvent",
    "StatusChangeEvent",
    "ErrorEvent",
    "CommandEvent",
    "StateSnapshotEvent",
    "PresenceEvent",
    "SchedulerModeEvent",
    "SchedulerDecisionEvent",
    "UsageResetEvent",
    "ConversationBlockEvent",
    "ConversationStateEvent",
    "TmuxFrameEvent",
    "BusEvent",
    "AgentEvent",
    "BUS_EVENT_ADAPTER",
    "EventFilter",
    "HelloBody",
    "SubArgs",
    "RpcArgs",
    "AckBody",
    "ErrBody",
    "WakeBody",
    "HelloMessage",
    "PubMessage",
    "SubMessage",
    "RpcMessage",
    "AckMessage",
    "ErrMessage",
    "WakeMessage",
    "WireMessage",
    "WIRE_MESSAGE_ADAPTER",
    "SOCKET_RUNTIME_SUBDIR",
    "SOCKET_BASENAME",
    "DEFAULT_RPC_TIMEOUT_S",
    "DEFAULT_HEARTBEAT_INTERVAL_S",
    "DEFAULT_LEASE_TTL_S",
    "DEFAULT_MAX_COMMAND_ATTEMPTS",
    "COMMAND_REAPER_INTERVAL_S",
    "PRESENCE_DISCONNECT_DEBOUNCE_S",
    "PRESENCE_USER_KINDS",
    "SUBSCRIBER_QUEUE_DEFAULT",
    "IDEMPOTENCY_WINDOW_S",
]
